### ISSUE TITLE
Gateway to use content-based redirection

### ISSUE THREAD

**Issue by jachro** [SwissDataScienceCenter/renku-gateway#169]
As a Renku owner, I'd like Gateway to redirect requests to `GET <renku>/<resource>` to `GET http://knowledge-graph/<resource>`. The redirect should happen only for `Accept: application/json`. Request with `Accept: text/html` should be redirected to the UI.

**Comment by ableuler** [comment]
Currently, this would have to happen at the ingress level. Calls without `/api` never make it to the gateway.

**Comment by jachro** [comment]
I see. Do you think it's possible to craft this kind of content-type based rules in Ingress?

**Comment by rokroskar** [comment]
@ableuler nothing is stopping us from sending more paths from the ingress to the gateway, right? I don't see any generic ingress rules that include content types.

**Comment by ableuler** [comment]
I looks like the [canary annotation](https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#canary) could do what we want, but it would definitely be an abuse of a feature which has a different purpose. Also, one can provide [custom configuration](https://kubernetes.github.io/ingress-nginx/examples/customization/custom-configuration/) as configMap to nginx ingress. I'm not sure but I guess this would also allow us to achieve what we want. 
The option mentioned by @rokroskar (routing all traffic to the UI through traefik) would of course also work, but it might be a bit backwards to have an ingress-controller deployed as DeamonSet which just forwards most requests to one or a couple of traefik pods inside the cluster. While this would most likely be no problem for all Renku instances that we'll be running in the foreseeable future, it would still be cleaner to just switch to traefik as ingress controller and provide the desired rules as configmap.
Otherwise, would it be an option to use `https://<hostname>/entities/projects/<namespace>/<project-name>` instead?

**Comment by jachro** [comment]
That's definitely an option to put some additional segment into the URL. It'd make them a bit longer but from the other side, we could also use the opportunity and add some versioning too. Maybe we can follow the GitLab API URL pattern `https://renku/api/v1/projects/<namespace>/<project-name>`?

**Comment by ableuler** [comment]
>  ...we could also use the opportunity and add some versioning too
As we've started without the version in the URL I'd prefer to go for [versioning through the header](https://restfulapi.net/versioning/). 
In a brief discussion, @rokroskar and I agreed that the human-readable web page URL `https://renku/projects/<namespace>/<project-name>` should serve as IDs for entities. Machine-readable information could be embedded in the HTML using the `<script type="application/ld+json">` tag.

**Comment by rokroskar** [comment]
But in addition to 

> the human-readable web page URL https://renku/projects/<namespace>/<project-name> should serve as IDs for entities. Rok suggested that machine-readable information could be embedded in the HTML using the <script type="application/ld+json"> tag.

we should/could provide api endpoints that serve just the json-ld, e.g. `https://renku/api/projects/<namespace>/<project>` or `https://renku/api/datasets/<uuid>`

**Comment by jachro** [comment]
Personally, I'm not convinced that HTML is more readable for a human being than JSON where you've got brief and concise info about the resource you're querying for. Especially, that the resource is used in `@id` of a json-ld object. I think people/machines who will be digesting json-ld objects and find out a resource, they could be much in favour of just taking it and passing to some curl/http client and get a similar type of response they took it from. Where else do you think people are going to see the resource urls? Are we displaying them in the UI? And I'd try to defend the `content-type` approach as for me that's the proper way of distinguishing the type of the response, not a different URL.

**Comment by rokroskar** [comment]
I think the point is that for entities that users interact with, e.g. `Project`, `Dataset`, the id in the KG of the entity should be a URI that if you paste it into the browser it takes you to a UI page about that entity. That's what we meant by "human readable" when @ableuler and I discussed this a few days ago. That also makes it easy for the UI to link entities when constructing different views - the UI doesn't have to know how the paths are constructed, it can just use the node IDs directly. Does that make sense?

**Comment by jachro** [comment]
Yes, but effectively as browser passes the `Accept` of `text/html` header it will work anyway as the single endpoint will return HTML.

**Comment by rokroskar** [comment]
This is an interesting read on the topic: https://www.w3.org/TR/cooluris/

One idea which we have not considered yet is to put the graph on its own subdomain, e.g. 

```https://renkulab.io/projects/<namespace>/<name>```

vs

```https://graph.renkulab.io/projects/<namespace>/<name>```

Similarly

```https://renkulab.io/projects/<namespace>/<name>/files/<path>/lineage```

would give me the UI with the graph lineage display and 

```https://graph.renkulab.io/projects/<namespace>/<name>/files/<path>/lineage```

would return the json-ld or json.

**Comment by ableuler** [comment]
> This is an interesting read on the topic: https://www.w3.org/TR/cooluris/

Ok, so the gist of it is that we should **not** use the URL where a page describing an entity (for example a project) can be found (say `https://renkulab.io/projects/<namespace>/<name>`) as `id` of that project in the knowledge graph, right? However, we still want `id`s that can be pasted into the browser.  
To achieve this (I don't really like the hash-URLs)  having a pattern like this
![image](https://user-images.githubusercontent.com/13014091/65495790-96afdd00-deb7-11e9-9d76-08be7620841d.png)
would seem quite natural to me, where in our case it would probably be `ld+json` instead of `rdf+xml` and `www.example.com/api/...` instead of `www.example.com/data/...`.

What I don't really understand is if in such a scenario, one would/should still embed JSON-LD annotations in the HTML page for search engines or is it enough to jus link to the alternative representation of the described entity like this once?
```
  <head>
    ...
    <link rel="alternate" type="application/ld+json"
          title="JSON-LD representation"
          href="https://renkulab.io/api/some/entity" />
  </head> ...
```

> The current URL at .../knowledge-graph seems clunky to me.

While I definitely agree that `graph.renkulab.io` or api.renkulab.io is nicer than`renkulab.io/knowledge-graph` or `renkulab.io/knowledge-graph`, this question should be specific to each deployment.

**Comment by jachro** [comment]
Personally, I quite like the alternative representation of the entity in the `<head>` as it seems to me to be simple and yet not make the HTML more complicated.

In the case of the redirects based on the content negotiation, it's more or less what I was leaning towards from the beginning. So I guess the endpoint which would get hit first would be some of our Traefik backed resources which would do then the redirect. Am I right?

In terms of the `knowledge-graph` path segment, I'd be more than happy to swap it with `api`, and it's a trivial change on the `knowledge-graph` service. However, by doing that we are kind of agreeing that all Renku APIs will be provided by that service, which maybe makes sense if KG is supposed to be the 'heart of Renku' (it's a bit poetic, isn't it?)

**Comment by ableuler** [comment]
> Personally, I quite like the alternative representation of the entity in the `<head>` as it seems to me to be simple and yet not make the HTML more complicated.

So it would mostly be search engines consuming that, right?

> In the case of the redirects based on the content negotiation, it's more or less what I was leaning towards from the beginning. So I guess the endpoint which would get hit first would be some of our Traefik backed resources which would do then the redirect. Am I right?

Traefik would seem the obvious place to do this to me too.

> In terms of the `knowledge-graph` path segment, I'd be more than happy to swap it with `api`, and it's a trivial change on the `knowledge-graph` service. However, by doing that we are kind of agreeing that all Renku APIs will be provided by that service, which maybe makes sense if KG is supposed to be the 'heart of Renku' (it's a bit poetic, isn't it?)

Yes, I completely agree. The core Renku resources are graph related and I'd prefer to have one api for all of Renku.

**Comment by ableuler** [comment]
Closing in favour of #174 and #175.

**Comment by rokroskar** [comment]
Reopening because the background info in this issue is useful - I'm turning it into an epic to group the related issues here.
